internal/db/merge.go

```diff
@@ -0,0 +1,361 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"container/list"
+	"context"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+
+	"github.com/sourcenetwork/corelog"
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/events"
+	"github.com/sourcenetwork/defradb/internal/core"
+	coreblock "github.com/sourcenetwork/defradb/internal/core/block"
+	"github.com/sourcenetwork/defradb/internal/db/base"
+	"github.com/sourcenetwork/defradb/internal/merkle/clock"
+	merklecrdt "github.com/sourcenetwork/defradb/internal/merkle/crdt"
+	"github.com/sourcenetwork/immutable"
+)
+
+func (db *db) handleMerges(ctx context.Context, merges events.Subscription[events.DAGMerge]) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case merge, ok := <-merges:
+			if !ok {
+				return
+			}
+			go func() {
+				err := db.executeMerge(ctx, merge)
+				if err != nil {
+					log.ErrorContextE(
+						ctx,
+						"Failed to execute merge",
+						err,
+						corelog.String("cid", merge.Cid.String()),
+					)
+				}
+			}()
+		}
+	}
+}
+
+func (db *db) executeMerge(ctx context.Context, dagMerge events.DAGMerge) error {
+	defer func() {
+		// Notify the caller that the merge is complete.
+		if dagMerge.MergeCompleteChan != nil {
+			close(dagMerge.MergeCompleteChan)
+		}
+	}()
+	ctx, txn, err := ensureContextTxn(ctx, db, false)
+	if err != nil {
+		return err
+	}
+	defer txn.Discard(ctx)
+	mp, err := db.newMergeProcessor(ctx, dagMerge.Cid, dagMerge.SchemaRoot)
+	if err != nil {
+		return err
+	}
+	mt, err := mp.getHeads(ctx)
```

Comment:
suggestion: The flow and concept-status of `mergeProcessor` could be improved I think.  At the moment it has the following problems:

1. It is only called in this function, and given that it's functions do not return anything at any stage, it is unclear why it is not wrapped up into a single func on `db`
2. Party because most of the functions do not return anything, it is unclear without reading the implementation of `mergeProcessor` as to whether the order in which each function is called matters or not, and if assuming it does matter, what that order needs to be.
3. It is strange that `getComposites` takes the return value of `getHeads`, when the host object is essentially a state-machine whose primary reason to be appears to manage state produced by it's host functions.
4. That `getComposites` does not return anything, despite being called `getFoo` confuses things further, and casts more doubt onto how many side affects each function has, making me want to read the implementation of `getHeads` to see if it also stores state in the state-machine as well as returning `mt`.
5. Partly because of the uncertainty around state-machineness and `mt`, and partly because it is quite dense and recursive code, I found it quite hard to be sure what `getComposites` was actually doing when reading it's implementation.

Overall I think `mergeProcessor` needs some work, although what kind of work is required I am not fully sure.  Maybe removing the semi-statemachineness and some more documentation would be enough.

Reply:
I agree with this. I pushed it as is because I was tired of all the troubleshooting I did but I will improve it before merge.